src/cpp/src/multinomial_decoding.cpp

```diff
@@ -0,0 +1,271 @@
+    const float inv_temp = 1.f / temp;
+    for (float* it = first; it != last; it++) {
+        *it *= inv_temp;
+    }
+}
+
+struct SamplingParameters {
+    int top_k;
+    float top_p;
+    float temperature;
+    float repetition_penalty;
+
+    SamplingParameters(ov::GenerationConfig generation_config) {
+        // parameters validation
+        OPENVINO_ASSERT(generation_config.top_k > 0,
+                        "top_k must be a strictly positive float, but got ",
+                        generation_config.top_p);
+        OPENVINO_ASSERT(generation_config.top_p > 0 || generation_config.top_p < 1.0f,
+                        "top_p must be a positive float > 0 and < 1, but got ",
+                        generation_config.top_p);
+        OPENVINO_ASSERT(generation_config.temperature > 0,
+                        "Temperature must be a strictly positive float, but got ",
+                        generation_config.temperature);
+        OPENVINO_ASSERT(generation_config.repetition_penalty > 0,
+                        "Repetition penalty must be a strictly positive float, but got ",
+                        generation_config.repetition_penalty);
+
+        top_k = generation_config.top_k;
+        top_p = generation_config.top_p;
+        temperature = generation_config.temperature;
+        repetition_penalty = generation_config.repetition_penalty;
+    }
+};
+
+struct RandomSampling {
+    SamplingParameters parameters;
+    RandomSampling(SamplingParameters parameters) : parameters{std::move(parameters)} {}
+
+    TokenIdScore get_out_token(float* logits, size_t vocab_size, std::vector<int64_t> tokens) {
+        // logits pre-process
+        if (parameters.repetition_penalty != 1.0f) {
+            sampling_repetition_penalty(logits, logits + vocab_size, tokens, parameters.repetition_penalty);
+        }
+
+        if (parameters.temperature > 0) {
+            sampling_temperature(logits, logits + vocab_size, parameters.temperature);
+        }
+
+        std::vector<TokenIdScore> token_scores(vocab_size);
+        for (int i = 0; i < vocab_size; i++) {
+            token_scores[i] = TokenIdScore(i, logits[i]);
+        }
+
+        // top_k sampling
+        if (0 < parameters.top_k && parameters.top_k < (int)token_scores.size()) {
+            sampling_top_k(token_scores.data(),
+                           token_scores.data() + parameters.top_k,
+                           token_scores.data() + token_scores.size());
+            token_scores.resize(parameters.top_k);
+        }
+
+        // top_p sampling
+        if (0.f < parameters.top_p && parameters.top_p < 1.0f) {
+            auto pos = sampling_top_p(token_scores.data(), token_scores.data() + token_scores.size(), parameters.top_p);
+            token_scores.resize(pos - token_scores.data());
+        }
+
+        // sample next token
+        sampling_softmax_inplace(token_scores.data(), token_scores.data() + token_scores.size());

```

Comment:
In `top_p` softmax applied to a copy vector and does not affect resulting distribution.
https://github.com/pavel-esir/openvino.genai/pull/6/files#diff-16dfb4a116b31e1dc48cb6d111de17ae77e0a0ec71f4000763e92cae56de34eeR53

Reply:
I see.
Will this save computing `pos` extra elements assuming `sample_top_p()` doesn't copy the vector?
```C++
if (0.f < top_p && top_p < 1.0f) {
    auto pos = sample_top_p(token_scores.data(), token_scores.data() + token_scores.size(), top_p);
    token_scores.resize(pos - token_scores.data());
} else {
    apply_softmax_inplace(token_scores)
}

// sample next token
for (size_t i = 0; i < token_scores.size(); i++) {
    logits[i] = token_scores[i].score;
}
```
Please, apply if so.

Reply:
I am a bit confused. `softmax` was not applied for `token_scores` after `sample_top_p` so we would need to apply it after.
The reason why it was not applied in `sample_top_p` is to keep distribution correct after resizing tokens scores vector.
Maybe I understood your comment wrong. If so let's sync offline.

Reply:
I see now.
1. The reported probability for the returned token isn't the one predicted by the model but a rescaled probability obtained after `top_k` and `p` truncation. Although it feels more natural to report a probability proved by a model. Is the current implementation aligned with transformers? If so, let it be.
2. There's `top_k` and p conflict. After `top_k` is applied, the distribution is rescaled before applying `top_p`. It's hard to build an intuition in mind to predict how `top_k` affects that distribution. The natural way would be to ask a user to provide `top_p` for model predicted distribution, but we corrupt it by applying `top_k`. Is the current implementation aligned with transformers? If so, let it be.
3. `sample_top_p()` has `vocab_size` var. The problem is that `vocab_size` is equal to `top_k` instead of being equal to model's vocab size. Can you rename it to `truncated_vocab_size` or something like this to reflect that it can be affected by `top_k`?

Reply:
1., 2. aligned with HF implementation
3. Renamed